crates/bevy_ecs/src/template.rs

```diff
@@ -0,0 +1,190 @@
+//! Functionality that relates to the [`Template`] trait.
+
+pub use bevy_ecs_macros::GetTemplate;
+
+use crate::{
+    bundle::Bundle,
+    entity::{Entity, EntityPath},
+    error::{BevyError, Result},
+    world::EntityWorldMut,
+};
+use alloc::{boxed::Box, vec, vec::Vec};
+use bevy_platform::collections::hash_map::Entry;
+use bevy_utils::TypeIdMap;
+use core::any::{Any, TypeId};
+use downcast_rs::{impl_downcast, Downcast};
+use variadics_please::all_tuples;
+
+/// A [`Template`] is something that, given a spawn context (target [`Entity`], [`World`](crate::world::World), etc), can produce a [`Template::Output`].
+pub trait Template {
+    /// The type of value produced by this [`Template`].
+    type Output;
+
+    /// Uses this template and the given `entity` context to produce a [`Template::Output`].
+    fn build(&mut self, entity: &mut EntityWorldMut) -> Result<Self::Output>;
```

Comment:
Just so I understand the broader design: if we have templates and scenes (which can contain multiple entities), why do we need bundle effects?

Reply:
BundleEffects still facilitate the `children![]` macro, usable in Bundle position. I expect that to fall out of favor as `bsn!` gets adoption, but `children![]` will still be useful for people that want to work with bundles.